mqtt-streaming/src/test/java/docs/javadsl/MqttSourceTest.java

```diff
@@ -0,0 +1,157 @@
+  }
+
+  @AfterClass
+  public static void teardown() {
+    TestKit.shutdownActorSystem(MqttSourceTest.system);
+  }
+
+  @After
+  public void assertStageStopping() {
+    //    StreamTestKit.assertAllStagesStopped(materializer);
+  }
+
+  @Test
+  public void atLeastOnceSubscription()
+      throws InterruptedException, ExecutionException, TimeoutException {
+    String testId = "1";
+    String clientId = "streaming/source-test/" + testId;
+    String topic = topicPrefix + testId;
+
+    List<Publish> received = new ArrayList<>();
+
+    // #at-least-once
+
+    Flow<Pair<Publish, MqttAckHandle>, Pair<Publish, MqttAckHandle>, NotUsed> businessLogic = // ???
+        // #at-least-once
+        Flow.<Pair<Publish, MqttAckHandle>>create()
+            .map(
+                pair -> {
+                  received.add(pair.first());
+                  return pair;
+                });
+
+    // #at-least-once
+
+    MqttClientSession mqttClientSession =
+        ActorMqttClientSession.create(MqttSessionSettings.create(), materializer, system);
+    MqttTcpTransportSettings transportSettings = MqttTcpTransportSettings.create("localhost");
+    MqttSubscribe subscriptions = MqttSubscriptions.atLeastOnce(topic);
+
+    Pair<
+            Pair<CompletionStage<List<Pair<String, ControlPacketFlags>>>, UniqueKillSwitch>,
+            CompletionStage<Done>>
+        stream =
+            MqttSource.atLeastOnce(
+                    mqttClientSession,
+                    transportSettings,
+                    MqttRestartSettings.create(),
+                    MqttConnectionSettings.create(clientId),
+                    subscriptions)
+                .via(businessLogic)
+                .mapAsync(
+                    1,
+                    pair -> {
+                      MqttAckHandle ackHandle = pair.second();
+                      return ackHandle.ack();
+                    })
+                .viaMat(KillSwitches.single(), Keep.both())
+                .toMat(Sink.ignore(), Keep.both())
+                .run(materializer);
+
+    CompletionStage<List<Pair<String, ControlPacketFlags>>> subscribed = stream.first().first();
+    UniqueKillSwitch killSwitch = stream.first().second();
+    CompletionStage<Done> streamCompletion = stream.second();
+    // #at-least-once
+
+    SourceQueueWithComplete<Command<?>> publishFlow =
+        //        // TODO remove hack
+        new akka.stream.impl.SourceQueueAdapter(
+            MqttSourceSpec.publish(
+                topic,
+                ControlPacketFlags.QoSExactlyOnceDelivery(),
+                transportSettings,
+                JavaConverters.asScalaBuffer(input).toList(),
+                materializer,
+                system,
+                executionContext));
+
+    Thread.sleep(2 * 1000);
+    // #at-least-once
+
+    // stop the subscription
+    killSwitch.shutdown();
+    // #at-least-once
+
+    // TODO this assert fails
+    //    assertThat(
```

Comment:
I have been fighting this a couple of hours, for some reason, the Java API doesn't really work :-(